src/Algebra/Graph.hs

```diff
@@ -1082,3 +1093,73 @@ sparsify graph = res
         m <- get
         put (m + 1)
         overlay <$> s `x` m <*> m `y` t
+
+{- Note [The rules of foldg]
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+The rules for foldg work like this.
+
+* Up to (but not including) phase 1, we use the "buildG/f" rule to
+  rewrite all saturated applications of f with its buildG/foldg
+  form, hoping for fusion to happen (through the "foldg/buildG" rule).
+
+  In phase 1 and 0, we switch off these rules, inline buildG, and
+  switch on the "graph/f" rule, which rewrites the "foldg/f"
+  thing back into plain map.
+
+  It's important that these two rules aren't both active at once
+  (along with build's unfolding) else we'd get an infinite loop
+  in the rules. Hence the activation control below.
+
+* FB functions are here to remember f after applying a "builG/f" rule.
+  These functions are higher-order functionsand therefore benefits from
+  inlining.
+
+* The "mapGFB" rule optimises compositions of mapG
```

Comment:
Where does the suffix `FB` come from? Perhaps from `foldr` and `build`? I think it is worth adding a remark about this.

Reply:
Or, maybe it's better to come up with a different suffix. I would just use `R` for all such intermediate functions, where `R` stands for "rewrite" and "rules" simultaneously :-)

Reply:
I don't know at all from were come these `FB` functions ^^. These names come from [`GHC.List`](http://hackage.haskell.org/package/base-4.11.0.0/docs/src/GHC.List.html#foldl).

I am not sure changing these suffix since they are common in GHC sources...

Reply:
The problem with the suffix `FB` is precisely that we both don't know what it means. And other readers of this code will most probably not know either :) So, why don't we make it easier for the readers to understand what's going on by using `R` and saying that throughout the Alga codebase the suffix `R` indicates functions used exclusively for implementing rewrite rules?